File: speedviewlib/src/main/java/com/github/anastr/speedviewlib/AwesomeSpeedometer.java
package com.github.anastr.speedviewlib;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RadialGradient;
import android.graphics.RectF;
import android.graphics.Shader;
import android.graphics.Typeface;
import android.util.AttributeSet;

import com.github.anastr.speedviewlib.components.Indicators.TriangleIndicator;

/**
 * this Library build By Anas Altair
 * see it on <a href="https://github.com/anastr/SpeedView">GitHub</a>
 */
public class AwesomeSpeedometer extends Speedometer {

    private Path markPath = new Path(),
            trianglesPath = new Path();
    private Paint markPaint = new Paint(Paint.ANTI_ALIAS_FLAG),
            ringPaint = new Paint(Paint.ANTI_ALIAS_FLAG),
            trianglesPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private RectF speedometerRect = new RectF();

    private int speedometerColor = 0xff00e6e6;

    public AwesomeSpeedometer(Context context) {
        this(context, null);
    }

    public AwesomeSpeedometer(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public AwesomeSpeedometer(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
        initAttributeSet(context, attrs);
    }

    @Override
    protected void defaultGaugeValues() {

        super.setTextColor(0xffffc260);
        super.setSpeedTextColor(0xFFFFFFFF);
        super.setUnitTextColor(0xFFFFFFFF);
        super.setTextTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD));
        super.setSpeedTextPosition(Position.CENTER);
        super.setUnitUnderSpeedText(true);
    }

    @Override
    protected void defaultSpeedometerValues() {
        super.setIndicator(new TriangleIndicator(getContext())
                .setIndicatorWidth(dpTOpx(25f))
                .setIndicatorColor(0xff00e6e6));
        super.setStartEndDegree(135, 135 + 320);
        super.setSpeedometerWidth(dpTOpx(60f));
        super.setBackgroundCircleColor(0xff212121);
        super.setTickNumber(9);
        super.setTickPadding(0);
    }

    private void init() {
        markPaint.setStyle(Paint.Style.STROKE);
        textPaint.setTextAlign(Paint.Align.CENTER);
        ringPaint.setStyle(Paint.Style.STROKE);
        trianglesPaint.setColor(0xff3949ab);
    }

    private void initAttributeSet(Context context, AttributeSet attrs) {
        if (attrs == null)
            return;
        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.AwesomeSpeedometer, 0, 0);

        speedometerColor = a.getColor(R.styleable.AwesomeSpeedometer_sv_speedometerColor, speedometerColor);
        trianglesPaint.setColor(a.getColor(R.styleable.AwesomeSpeedometer_sv_trianglesColor, trianglesPaint.getColor()));
        a.recycle();
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldW, int oldH) {
        super.onSizeChanged(w, h, oldW, oldH);

        updateGradient();
        updateBackgroundBitmap();
    }

    private void updateGradient() {
        float stop = (getSizePa() *.5f - getSpeedometerWidth()) / (getSizePa() *.5f);
        float stop2 = stop+((1f-stop)*.1f);
        float stop3 = stop+((1f-stop)*.36f);
        float stop4 = stop+((1f-stop)*.64f);
        float stop5 = stop+((1f-stop)*.9f);
        int []colors = new int[]{getBackgroundCircleColor(), speedometerColor, getBackgroundCircleColor()
                , getBackgroundCircleColor(), speedometerColor, speedometerColor};
        Shader radialGradient = new RadialGradient(getSize() *.5f, getSize() *.5f, getSizePa() *.5f
                , colors, new float[]{stop, stop2, stop3, stop4, stop5, 1f}, Shader.TileMode.CLAMP);
        ringPaint.setShader(radialGradient);
    }

    private void initDraw() {
        ringPaint.setStrokeWidth(getSpeedometerWidth());
        markPaint.setColor(getMarkColor());
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        drawSpeedUnitText(canvas);
        drawIndicator(canvas);
        drawNotes(canvas);
    }

    @Override
    protected void updateBackgroundBitmap() {
        Canvas c = createBackgroundBitmapCanvas();
        initDraw();

        float markH = getViewSizePa()/22f;
        markPath.reset();
        markPath.moveTo(getSize() *.5f, getPadding());
        markPath.lineTo(getSize() *.5f, markH + getPadding());
        markPaint.setStrokeWidth(markH/5f);

        float triangleHeight = getViewSizePa() / 20f;
        setInitTickPadding(triangleHeight);

        trianglesPath.reset();
        trianglesPath.moveTo(getSize() *.5f, getPadding() + getViewSizePa()/20f);
        float triangleWidth = getViewSize()/20f;
        trianglesPath.lineTo(getSize() *.5f - triangleWidth/2f, getPadding());
        trianglesPath.lineTo(getSize() *.5f + triangleWidth/2f, getPadding());

        float risk = getSpeedometerWidth() *.5f + getPadding();
        speedometerRect.set(risk, risk, getSize() -risk, getSize() -risk);
        c.drawArc(speedometerRect, 0f, 360f, false, ringPaint);

        drawMarks(c);
        drawTicks(c);
    }

    protected void drawMarks(Canvas c) {
        for (int i=0; i < getTickNumber(); i++) {
            float d = getDegreeAtSpeed(getTicks().get(i)) + 90f;
            c.save();
            c.rotate(d, getSize() *.5f, getSize() *.5f);

            c.drawPath(trianglesPath, trianglesPaint);
            if(i+1 != getTickNumber()) {
                c.save();
                float d2 = getDegreeAtSpeed(getTicks().get(i + 1)) + 90f;
                float eachDegree = d2 - d;
                for (int j = 1; j < 10; j++) {
                    c.rotate(eachDegree * .1f, getSize() * .5f, getSize() * .5f);
                    if (j == 5)
                        markPaint.setStrokeWidth(getSize() / 22f / 5);
                    else
                        markPaint.setStrokeWidth(getSize() / 22f / 9);
                    c.drawPath(markPath, markPaint);
                }
                c.restore();
            }
            c.restore();
        }
    }

    @Override
    public void setSpeedometerWidth(float speedometerWidth) {
        super.setSpeedometerWidth(speedometerWidth);
        float risk = speedometerWidth *.5f;
        speedometerRect.set(risk, risk, getSize() -risk, getSize() -risk);
        updateGradient();
        updateBackgroundBitmap();
        invalidate();
    }

    public int getSpeedometerColor() {
        return speedometerColor;
    }

    public void setSpeedometerColor(int speedometerColor) {
        this.speedometerColor = speedometerColor;
        updateGradient();
        updateBackgroundBitmap();
        invalidate();
    }

    public int getTrianglesColor() {
        return trianglesPaint.getColor();
    }

    public void setTrianglesColor(int trianglesColor) {
        trianglesPaint.setColor(trianglesColor);
        updateBackgroundBitmap();
        invalidate();
    }

    /**
     * this Speedometer doesn't use this method.
     * @return {@code Color.TRANSPARENT} always.
     */
    @Deprecated
    @Override
    public int getLowSpeedColor() {
        return 0;
    }

    /**
     * this Speedometer doesn't use this method.
     * @param lowSpeedColor nothing.
     */
    @Deprecated
    @Override
    public void setLowSpeedColor(int lowSpeedColor) {
    }

    /**
     * this Speedometer doesn't use this method.
     * @return {@code Color.TRANSPARENT} always.
     */
    @Deprecated
    @Override
    public int getMediumSpeedColor() {
        return 0;
    }

    /**
     * this Speedometer doesn't use this method.
     * @param mediumSpeedColor nothing.
     */
    @Deprecated
    @Override
    public void setMediumSpeedColor(int mediumSpeedColor) {
    }

    /**
     * this Speedometer doesn't use this method.
     * @return {@code Color.TRANSPARENT} always.
     */
    @Deprecated
    @Override
    public int getHighSpeedColor() {
        return 0;
    }

    /**
     * this Speedometer doesn't use this method.
     * @param highSpeedColor nothing.
     */
    @Deprecated
    @Override
    public void setHighSpeedColor(int highSpeedColor) {
    }
}

File: speedviewlib/src/main/java/com/github/anastr/speedviewlib/AwesomeSpeedometer.kt
package com.github.anastr.speedviewlib

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import com.github.anastr.speedviewlib.components.indicators.TriangleIndicator

/**
 * this Library build By Anas Altair
 * see it on [GitHub](https://github.com/anastr/SpeedView)
 */
open class AwesomeSpeedometer @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0,
) : Speedometer(context, attrs, defStyleAttr) {

    private val markPath = Path()
    private val trianglesPath = Path()
    private val customMarkPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val ringPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val trianglesPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val speedometerRect = RectF()

    private var speedometerColor = 0xff00e6e6.toInt()

    var trianglesColor: Int
        get() = trianglesPaint.color
        set(trianglesColor) {
            trianglesPaint.color = trianglesColor
            invalidateGauge()
        }

    override var speedometerWidth
        get() = super.speedometerWidth
        set(speedometerWidth) {
            super.speedometerWidth = speedometerWidth
            // in case AwesomeSpeedometer not initialized
            if (speedometerRect != null) {
                val risk = speedometerWidth * .5f
                speedometerRect.set(risk, risk, size - risk, size - risk)
                updateGradient()
                invalidateGauge()
            }
        }

    init {
        init()
        initAttributeSet(context, attrs)
    }

    override fun defaultGaugeValues() {

        super.speedometerWidth = dpTOpx(60f)
        super.textColor = 0xffffc260.toInt()
        super.speedTextColor = 0xFFFFFFFF.toInt()
        super.unitTextColor = 0xFFFFFFFF.toInt()
        super.textTypeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
        super.speedTextPosition = Position.CENTER
        super.unitUnderSpeedText = true
    }

    override fun defaultSpeedometerValues() {
        indicator = TriangleIndicator(context)
        indicator.apply {
            width = dpTOpx(25f)
            color = 0xff00e6e6.toInt()
        }
        super.setStartEndDegree(135, 135 + 320)
        super.backgroundCircleColor = 0xff212121.toInt()
        super.tickNumber = 9
        super.tickPadding = 0f
    }

    private fun init() {
        customMarkPaint.style = Paint.Style.STROKE
        textPaint.textAlign = Paint.Align.CENTER
        ringPaint.style = Paint.Style.STROKE
        trianglesPaint.color = 0xff3949ab.toInt()
    }

    private fun initAttributeSet(context: Context, attrs: AttributeSet?) {
        if (attrs == null)
            return
        val a = context.theme.obtainStyledAttributes(attrs, R.styleable.AwesomeSpeedometer, 0, 0)

        speedometerColor = a.getColor(R.styleable.AwesomeSpeedometer_sv_speedometerColor, speedometerColor)
        trianglesPaint.color = a.getColor(R.styleable.AwesomeSpeedometer_sv_trianglesColor, trianglesPaint.color)
        a.recycle()
    }

    override fun onSizeChanged(w: Int, h: Int, oldW: Int, oldH: Int) {
        super.onSizeChanged(w, h, oldW, oldH)

        updateGradient()
        updateBackgroundBitmap()
    }

    private fun updateGradient() {
        val stop = (sizePa * .5f - speedometerWidth) / (sizePa * .5f)
        val stop2 = stop + (1f - stop) * .1f
        val stop3 = stop + (1f - stop) * .36f
        val stop4 = stop + (1f - stop) * .64f
        val stop5 = stop + (1f - stop) * .9f
        val colors = intArrayOf(backgroundCircleColor, speedometerColor, backgroundCircleColor, backgroundCircleColor, speedometerColor, speedometerColor)
        val radialGradient = RadialGradient(size * .5f, size * .5f, sizePa * .5f, colors, floatArrayOf(stop, stop2, stop3, stop4, stop5, 1f), Shader.TileMode.CLAMP)
        ringPaint.shader = radialGradient
    }

    private fun initDraw() {
        ringPaint.strokeWidth = speedometerWidth
        customMarkPaint.color = markColor
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        drawSpeedUnitText(canvas)
        drawIndicator(canvas)
        drawNotes(canvas)
    }

    override fun updateBackgroundBitmap() {
        val c = createBackgroundBitmapCanvas()
        initDraw()

        val markH = viewSizePa / 22f
        markPath.reset()
        markPath.moveTo(size * .5f, padding.toFloat())
        markPath.lineTo(size * .5f, markH + padding)
        customMarkPaint.strokeWidth = markH / 5f

        val triangleHeight = viewSizePa / 20f
        initTickPadding = triangleHeight

        trianglesPath.reset()
        trianglesPath.moveTo(size * .5f, padding + viewSizePa / 20f)
        val triangleWidth = viewSize / 20f
        trianglesPath.lineTo(size * .5f - triangleWidth / 2f, padding.toFloat())
        trianglesPath.lineTo(size * .5f + triangleWidth / 2f, padding.toFloat())

        val risk = speedometerWidth * .5f + padding
        speedometerRect.set(risk, risk, size - risk, size - risk)
        c.drawArc(speedometerRect, 0f, 360f, false, ringPaint)

        drawCustomMarks(c)
        drawMarks(c)
        drawTicks(c)
    }

    protected fun drawCustomMarks(c: Canvas) {
        val range = getEndDegree() - getStartDegree()
        ticks.forEachIndexed { index, t ->
            val d = getStartDegree() + range * t
            c.save()
            c.rotate(d + 90f, size * .5f, size * .5f)

            c.drawPath(trianglesPath, trianglesPaint)
            if (index + 1 != tickNumber) {
                c.save()
                val d2 = getStartDegree() + range * ticks[index + 1]
                val eachDegree = d2 - d
                for (j in 1..9) {
                    c.rotate(eachDegree * .1f, size * .5f, size * .5f)
                    if (j == 5)
                        customMarkPaint.strokeWidth = size.toFloat() / 22f / 5f
                    else
                        customMarkPaint.strokeWidth = size.toFloat() / 22f / 9f
                    c.drawPath(markPath, customMarkPaint)
                }
                c.restore()
            }
            c.restore()
        }
    }

    fun getSpeedometerColor(): Int {
        return speedometerColor
    }

    fun setSpeedometerColor(speedometerColor: Int) {
        this.speedometerColor = speedometerColor
        updateGradient()
        invalidateGauge()
    }
}
